File: src/lib/types/flow.ts
import type { SerializedScene } from '../../routes/tools/oled/types';

// ==================== Flow Node Types ====================

export type FlowNodeType = 'intro' | 'home' | 'menu' | 'submenu' | 'custom' | 'screensaver';

export interface FlowVariable {
	name: string;
	type: 'int' | 'int8' | 'int16' | 'int32';
	defaultValue: number;
	persist: boolean;
	min?: number;
	max?: number;
}

export interface WidgetPlacement {
	widgetId: string;
	x: number;
	y: number;
	config: Record<string, unknown>;
	boundVariable?: string;
}

export interface FlowNode {
	id: string;
	type: FlowNodeType;
	label: string;
	position: { x: number; y: number };
	gpcCode: string;
	oledScene: SerializedScene | null;
	oledWidgets: WidgetPlacement[];
	comboCode: string;
	isInitialState: boolean;
	variables: FlowVariable[];
	onEnter: string;
	onExit: string;
	chunkRef: string | null;
}

// ==================== Flow Edge Types ====================

export type FlowConditionType =
	| 'button_press'
	| 'button_hold'
	| 'timeout'
	| 'variable'
	| 'custom';

export interface FlowCondition {
	type: FlowConditionType;
	button?: string;
	timeoutMs?: number;
	variable?: string;
	comparison?: '==' | '!=' | '>' | '<' | '>=' | '<=';
	value?: number;
	customCode?: string;
}

export interface FlowEdge {
	id: string;
	sourceNodeId: string;
	targetNodeId: string;
	sourcePort: string;
	targetPort: string;
	label: string;
	condition: FlowCondition;
}

// ==================== Flow Graph ====================

export interface FlowSettings {
	screenTimeoutMs: number;
	defaultFont: string;
	includeCommonUtils: boolean;
	persistenceEnabled: boolean;
	buttonMapping: {
		confirm: string;
		cancel: string;
		up: string;
		down: string;
		left: string;
		right: string;
	};
}

export interface FlowGraph {
	id: string;
	name: string;
	version: number;
	nodes: FlowNode[];
	edges: FlowEdge[];
	globalVariables: FlowVariable[];
	globalCode: string;
	settings: FlowSettings;
	createdAt: number;
	updatedAt: number;
}

// ==================== Chunks ====================

export interface ChunkEdgeTemplate {
	label: string;
	condition: FlowCondition;
	direction: 'outgoing' | 'incoming';
}

export interface ChunkParameter {
	key: string;
	label: string;
	type: 'string' | 'number' | 'boolean' | 'button' | 'code';
	defaultValue: string;
	description: string;
}

export interface FlowChunk {
	id: string;
	name: string;
	description: string;
	category: string;
	tags: string[];
	nodeTemplate: Partial<FlowNode>;
	edgeTemplates: ChunkEdgeTemplate[];
	parameters: ChunkParameter[];
	createdAt: number;
	updatedAt: number;
}

// ==================== OLED Widget Types ====================

export interface WidgetParam {
	key: string;
	label: string;
	type: 'number' | 'boolean' | 'select';
	default: unknown;
	min?: number;
	max?: number;
	options?: { value: unknown; label: string }[];
}

export interface OledWidgetDef {
	id: string;
	name: string;
	category: 'bar' | 'indicator' | 'diagnostic' | 'decorative';
	description: string;
	width: number;
	height: number;
	params: WidgetParam[];
	render: (config: Record<string, unknown>, value: number, pixels: Uint8Array, x: number, y: number) => void;
	generateGpc: (config: Record<string, unknown>, varName: string, x: number, y: number) => string;
}

// ==================== Template Types ====================

export interface TemplateParam {
	key: string;
	label: string;
	type: 'string' | 'number' | 'boolean' | 'select';
	default: unknown;
	options?: { value: unknown; label: string }[];
	description: string;
}

export interface ScriptTemplate {
	id: string;
	name: string;
	description: string;
	category: 'state-machine' | 'oled-menu' | 'persistence' | 'screensaver' | 'utility';
	tags: string[];
	generatesCode: boolean;
	generatesFlowGraph: boolean;
	params: TemplateParam[];
	generateCode?: (params: Record<string, unknown>) => string;
	generateFlowGraph?: (params: Record<string, unknown>) => FlowGraph;
}

// ==================== Defaults / Helpers ====================

export const DEFAULT_FLOW_SETTINGS: FlowSettings = {
	screenTimeoutMs: 5000,
	defaultFont: 'OLED_FONT_SMALL',
	includeCommonUtils: true,
	persistenceEnabled: true,
	buttonMapping: {
		confirm: 'CONFIRM_BTN',
		cancel: 'CANCEL_BTN',
		up: 'UP_BTN',
		down: 'DOWN_BTN',
		left: 'LEFT_BTN',
		right: 'RIGHT_BTN',
	},
};

export function createEmptyFlowGraph(name: string): FlowGraph {
	return {
		id: crypto.randomUUID(),
		name,
		version: 1,
		nodes: [],
		edges: [],
		globalVariables: [],
		globalCode: '',
		settings: { ...DEFAULT_FLOW_SETTINGS },
		createdAt: Date.now(),
		updatedAt: Date.now(),
	};
}

export function createFlowNode(
	type: FlowNodeType,
	label: string,
	position: { x: number; y: number }
): FlowNode {
	return {
		id: crypto.randomUUID(),
		type,
		label,
		position,
		gpcCode: '',
		oledScene: null,
		oledWidgets: [],
		comboCode: '',
		isInitialState: false,
		variables: [],
		onEnter: '',
		onExit: '',
		chunkRef: null,
	};
}

export function createFlowEdge(
	sourceNodeId: string,
	targetNodeId: string,
	label: string,
	condition: FlowCondition
): FlowEdge {
	return {
		id: crypto.randomUUID(),
		sourceNodeId,
		targetNodeId,
		sourcePort: 'right',
		targetPort: 'left',
		label,
		condition,
	};
}

export const NODE_COLORS: Record<FlowNodeType, string> = {
	intro: '#3b82f6',
	home: '#22c55e',
	menu: '#a855f7',
	submenu: '#f97316',
	custom: '#6b7280',
	screensaver: '#06b6d4',
};

export const NODE_LABELS: Record<FlowNodeType, string> = {
	intro: 'Intro',
	home: 'Home',
	menu: 'Menu',
	submenu: 'Submenu',
	custom: 'Custom',
	screensaver: 'Screensaver',
};

File: src-tauri/src/models/flow.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Complete flow graph definition stored as flow.json in game directories
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FlowGraph {
    pub id: String,
    pub name: String,
    pub version: u32,
    pub nodes: Vec<FlowNode>,
    pub edges: Vec<FlowEdge>,
    #[serde(default)]
    pub global_variables: Vec<FlowVariable>,
    #[serde(default)]
    pub global_code: String,
    pub settings: FlowSettings,
    #[serde(default)]
    pub created_at: u64,
    #[serde(default)]
    pub updated_at: u64,
}

/// A node in the flow graph representing a state/screen
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FlowNode {
    pub id: String,
    pub r#type: String, // intro, home, menu, submenu, custom, screensaver
    pub label: String,
    pub position: Position,
    #[serde(default)]
    pub gpc_code: String,
    #[serde(default)]
    pub oled_scene: Option<SerializedScene>,
    #[serde(default)]
    pub oled_widgets: Vec<WidgetPlacement>,
    #[serde(default)]
    pub combo_code: String,
    #[serde(default)]
    pub is_initial_state: bool,
    #[serde(default)]
    pub variables: Vec<FlowVariable>,
    #[serde(default)]
    pub on_enter: String,
    #[serde(default)]
    pub on_exit: String,
    #[serde(default)]
    pub chunk_ref: Option<String>,
}

/// An edge connecting two nodes representing a transition
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FlowEdge {
    pub id: String,
    pub source_node_id: String,
    pub target_node_id: String,
    #[serde(default = "default_port")]
    pub source_port: String,
    #[serde(default = "default_port")]
    pub target_port: String,
    #[serde(default)]
    pub label: String,
    pub condition: FlowCondition,
}

fn default_port() -> String {
    "right".to_string()
}

/// Transition condition for an edge
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FlowCondition {
    pub r#type: String, // button_press, button_hold, timeout, variable, custom
    #[serde(default)]
    pub button: Option<String>,
    #[serde(default)]
    pub timeout_ms: Option<u32>,
    #[serde(default)]
    pub variable: Option<String>,
    #[serde(default)]
    pub comparison: Option<String>,
    #[serde(default)]
    pub value: Option<i32>,
    #[serde(default)]
    pub custom_code: Option<String>,
}

/// A variable used in the flow
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FlowVariable {
    pub name: String,
    pub r#type: String, // int, int8, int16, int32
    #[serde(default)]
    pub default_value: i32,
    #[serde(default)]
    pub persist: bool,
    #[serde(default)]
    pub min: Option<i32>,
    #[serde(default)]
    pub max: Option<i32>,
}

/// OLED scene data (matches frontend SerializedScene)
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SerializedScene {
    pub id: String,
    pub name: String,
    pub pixels: String, // Base64-encoded Uint8Array
}

/// Widget placement within a node's OLED area
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WidgetPlacement {
    pub widget_id: String,
    pub x: i32,
    pub y: i32,
    #[serde(default)]
    pub config: HashMap<String, serde_json::Value>,
    #[serde(default)]
    pub bound_variable: Option<String>,
}

/// Position on the canvas
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Position {
    pub x: f64,
    pub y: f64,
}

/// Flow settings
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FlowSettings {
    #[serde(default = "default_timeout")]
    pub screen_timeout_ms: u32,
    #[serde(default = "default_font")]
    pub default_font: String,
    #[serde(default = "default_true")]
    pub include_common_utils: bool,
    #[serde(default = "default_true")]
    pub persistence_enabled: bool,
    #[serde(default)]
    pub button_mapping: FlowButtonMapping,
}

fn default_timeout() -> u32 {
    5000
}
fn default_font() -> String {
    "OLED_FONT_SMALL".to_string()
}
fn default_true() -> bool {
    true
}

/// Button mapping for flow navigation
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FlowButtonMapping {
    #[serde(default = "default_confirm")]
    pub confirm: String,
    #[serde(default = "default_cancel")]
    pub cancel: String,
    #[serde(default = "default_up")]
    pub up: String,
    #[serde(default = "default_down")]
    pub down: String,
    #[serde(default = "default_left")]
    pub left: String,
    #[serde(default = "default_right")]
    pub right: String,
}

impl Default for FlowButtonMapping {
    fn default() -> Self {
        Self {
            confirm: default_confirm(),
            cancel: default_cancel(),
            up: default_up(),
            down: default_down(),
            left: default_left(),
            right: default_right(),
        }
    }
}

fn default_confirm() -> String {
    "CONFIRM_BTN".to_string()
}
fn default_cancel() -> String {
    "CANCEL_BTN".to_string()
}
fn default_up() -> String {
    "UP_BTN".to_string()
}
fn default_down() -> String {
    "DOWN_BTN".to_string()
}
fn default_left() -> String {
    "LEFT_BTN".to_string()
}
fn default_right() -> String {
    "RIGHT_BTN".to_string()
}

/// Reusable chunk stored in workspace chunks/ directory
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FlowChunk {
    pub id: String,
    pub name: String,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub category: String,
    #[serde(default)]
    pub tags: Vec<String>,
    pub node_template: serde_json::Value, // Partial FlowNode as JSON
    #[serde(default)]
    pub edge_templates: Vec<ChunkEdgeTemplate>,
    #[serde(default)]
    pub parameters: Vec<ChunkParameter>,
    #[serde(default)]
    pub created_at: u64,
    #[serde(default)]
    pub updated_at: u64,
}

/// Edge template for a chunk
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChunkEdgeTemplate {
    pub label: String,
    pub condition: FlowCondition,
    pub direction: String, // "outgoing" | "incoming"
}

/// Configurable parameter for a chunk
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChunkParameter {
    pub key: String,
    pub label: String,
    pub r#type: String, // string, number, boolean, button, code
    pub default_value: String,
    #[serde(default)]
    pub description: String,
}
